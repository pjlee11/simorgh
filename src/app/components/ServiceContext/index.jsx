import React from 'react';
import { func, node, string } from 'prop-types';
import services from '../../lib/config/services';

/* Create ServiceContext using the default service */
export const ServiceContext = React.createContext(services.default);

export const ServiceContextProvider = ({ children, service }) => (
  <ServiceContext.Provider value={services[service]}>
    {children}
  </ServiceContext.Provider>
);

export const ServiceContextConsumer = ({ children }) => (
  <ServiceContext.Consumer>{children}</ServiceContext.Consumer>
);

ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string,
};

ServiceContextProvider.defaultProps = {
  service: 'default',
};

ServiceContextConsumer.propTypes = {
  children: func.isRequired,
};
