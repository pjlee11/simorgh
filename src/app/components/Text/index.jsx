import React from 'react';
import { string, node, func } from 'prop-types';
import Markdown from 'markdown-to-jsx';
import InlineLink from '../../containers/InlineLink';
import Paragraph from '../Paragraph';

const Italic = ({ children }) => <i>{children}</i>;
const Bold = ({ children }) => <b>{children}</b>;

const Text = ({ text, paragraphOverride }) => {
  if (!text) return null;

  const options = {
    forceBlock: true,
    overrides: {
      p: {
        component: paragraphOverride,
      },
      em: {
        component: Italic,
      },
      strong: {
        component: Bold,
      },
      a: {
        component: InlineLink,
      },
    },
  };

  return <Markdown options={options}>{text}</Markdown>;
};

Italic.propTypes = {
  children: node.isRequired,
};

Bold.propTypes = {
  children: node.isRequired,
};

Text.propTypes = {
  text: string.isRequired,
  paragraphOverride: func,
};

Text.defaultProps = {
  paragraphOverride: Paragraph,
};

export default Text;
