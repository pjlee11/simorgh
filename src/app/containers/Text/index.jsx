import React from 'react';
import { string } from 'prop-types';
import paragraph from '../Paragraph';
import Blocks from '../Blocks';
import { textModelPropTypes } from '../../models/propTypes/text';

const componentsToRender = { paragraph };

const shouldPrependHR = typeOfPreviousBlock => typeOfPreviousBlock === 'text';

const TextContainer = ({ blocks, typeOfPreviousBlock }) => {
  if (!blocks) return null;

  const HorizontalRule = shouldPrependHR(typeOfPreviousBlock) ? <hr /> : null;

  return (
    <React.Fragment>
      {HorizontalRule}
      <Blocks blocks={blocks} componentsToRender={componentsToRender} />
    </React.Fragment>
  );
};

TextContainer.propTypes = {
  ...textModelPropTypes,
  typeOfPreviousBlock: string,
};

TextContainer.defaultProps = {
  typeOfPreviousBlock: null,
};

export default TextContainer;
