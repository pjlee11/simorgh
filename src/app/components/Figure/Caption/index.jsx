import React from 'react';
import { node } from 'prop-types';
import styled from 'styled-components';
import {
  FF_NEWS_SANS_REG,
  C_STORM,
  GEL_SPACING,
  GEL_SPACING_DBL,
  C_STONE,
} from '../../../lib/constants/styles';
import VisuallyHiddenText from '../../VisuallyHiddenText';
import mediaQuery from '../../../helpers/mediaQueries';
import { T_LONG_PRIMER } from '../../../lib/constants/typography';
import { ServiceContext } from '../../ServiceContext';

const StyledCaption = styled.figcaption`
  ${T_LONG_PRIMER};
  background-color: ${C_STONE};
  color: ${C_STORM};
  font-family: ${FF_NEWS_SANS_REG};
  padding: ${GEL_SPACING};
  width: 100%;
  ${mediaQuery.laptopAndLarger} {
    padding: ${GEL_SPACING} ${GEL_SPACING_DBL};
  }
`;

const Caption = ({ children }) => (
  <StyledCaption>
    <ServiceContext.Consumer>
      {({ imageCaptionOffscreenText }) => (
        <VisuallyHiddenText>{imageCaptionOffscreenText}</VisuallyHiddenText>
      )}
    </ServiceContext.Consumer>
    {children}
  </StyledCaption>
);

Caption.propTypes = {
  // children will be "element.isRequired" in the future to support embedded <lang> and markdown
  children: node.isRequired,
};

export default Caption;
