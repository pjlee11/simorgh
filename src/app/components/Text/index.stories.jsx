import React from 'react';
import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
import Text from './index';

storiesOf('Text', module)
  .add('default', () => <Text text="This is text with no formatting." />)
  .add('bold', () => (
    <Text text="This is text with **some bold formatting**." />
  ))
  .add('italic', () => (
    <Text text="This is text with _some italic formatting_." />
  ))
  .add('strike-through', () => (
    <Text text="This is text with ~~some strike-through formatting~~." />
  ))
  .add('inline code', () => (
    <Text text="This is text with `some inline code`." />
  ))
  .add('inline link', () => (
    <Text text="This is text that contains an [inline link](https://www.bbc.com/news) inside it." />
  ));
