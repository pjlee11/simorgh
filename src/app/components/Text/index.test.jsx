import React from 'react';
import { shouldMatchSnapshot, isNull } from '../../helpers/tests/testHelpers';
import Text from './index';
import Caption from '../Figure/Caption/index';
import { ServiceContext } from '../ServiceContext';

describe('Text', () => {
  describe('with no data', () => {
    isNull('should return null', <Text />);
  });

  shouldMatchSnapshot(
    'should render bold',
    <Text text="This is text with **some bold formatting**." />,
  );

  shouldMatchSnapshot(
    'should render italics',
    <Text text="This is text with _some italic formatting_." />,
  );

  shouldMatchSnapshot(
    'should render strike-through',
    <Text text="This is text with ~~some strike-through formatting~~." />,
  );

  shouldMatchSnapshot(
    'should render inline-code',
    <Text text="This is text with `some inline code`." />,
  );

  shouldMatchSnapshot(
    'should render an inline link',
    <Text text="This is text that contains an [inline link](https://www.bbc.com/news) inside it." />,
  );

  const serviceContextStubNews = {
    imageCaptionOffscreenText: 'Image caption, ',
  };

  shouldMatchSnapshot(
    'should render a <figcaption> as per the Caption component override',
    <ServiceContext.Provider value={serviceContextStubNews}>
      <Text
        text="This is text that contains an [inline link](https://www.bbc.com/news) inside it."
        paragraphOverride={Caption}
      />
    </ServiceContext.Provider>,
  );
});
