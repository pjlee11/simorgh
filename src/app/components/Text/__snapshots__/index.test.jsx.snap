// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should render a <figcaption> as per the Caption component override 1`] = `
.c2 {
  color: #0F556C;
  padding: 2px;
  border-bottom: 1px solid #0F556C;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:visited {
  color: #5C5752;
  border-bottom: 1px solid #5C5752;
}

.c2:focus {
  background-color: #0F556C;
  border-bottom: 1px solid #F5F3F1;
  color: #F5F3F1;
}

.c2:hover {
  color: #0F556C;
  background-color: #C3DEE7;
  border-bottom: 1px solid #0F556C;
}

.c1 {
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  top: 0;
}

.c0 {
  font-size: 0.9375em;
  line-height: 1.125rem;
  background-color: #D5D0CD;
  color: #404040;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  padding: 0.5rem;
  width: 100%;
}

@media (min-width:37.5em) {
  .c0 {
    font-size: 0.875em;
  }
}

@media (min-width:37.5em) {
  .c0 {
    padding: 0.5rem 1rem;
  }
}

<figcaption
  className="c0"
>
  <span
    className="c1"
  >
    Image caption, 
  </span>
  This is text that contains an 
  <a
    className="c2"
    href="https://www.bbc.com/news"
  >
    inline link
  </a>
   inside it.
</figcaption>
`;

exports[`Text should render an inline link 1`] = `
.c1 {
  color: #0F556C;
  padding: 2px;
  border-bottom: 1px solid #0F556C;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:visited {
  color: #5C5752;
  border-bottom: 1px solid #5C5752;
}

.c1:focus {
  background-color: #0F556C;
  border-bottom: 1px solid #F5F3F1;
  color: #F5F3F1;
}

.c1:hover {
  color: #0F556C;
  background-color: #C3DEE7;
  border-bottom: 1px solid #0F556C;
}

.c0 {
  color: #404040;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  padding-bottom: 1rem;
  margin: 0;
  font-size: 0.9375em;
  line-height: 1.25rem;
}

@media (min-width:20em) {
  .c0 {
    font-size: 1em;
    line-height: 1.375rem;
  }
}

<p
  className="c0"
>
  This is text that contains an 
  <a
    className="c1"
    href="https://www.bbc.com/news"
  >
    inline link
  </a>
   inside it.
</p>
`;

exports[`Text should render bold 1`] = `
.c0 {
  color: #404040;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  padding-bottom: 1rem;
  margin: 0;
  font-size: 0.9375em;
  line-height: 1.25rem;
}

@media (min-width:20em) {
  .c0 {
    font-size: 1em;
    line-height: 1.375rem;
  }
}

<p
  className="c0"
>
  This is text with 
  <b>
    some bold formatting
  </b>
  .
</p>
`;

exports[`Text should render inline-code 1`] = `
.c0 {
  color: #404040;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  padding-bottom: 1rem;
  margin: 0;
  font-size: 0.9375em;
  line-height: 1.25rem;
}

@media (min-width:20em) {
  .c0 {
    font-size: 1em;
    line-height: 1.375rem;
  }
}

<p
  className="c0"
>
  This is text with 
  <code>
    some inline code
  </code>
  .
</p>
`;

exports[`Text should render italics 1`] = `
.c0 {
  color: #404040;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  padding-bottom: 1rem;
  margin: 0;
  font-size: 0.9375em;
  line-height: 1.25rem;
}

@media (min-width:20em) {
  .c0 {
    font-size: 1em;
    line-height: 1.375rem;
  }
}

<p
  className="c0"
>
  This is text with 
  <i>
    some italic formatting
  </i>
  .
</p>
`;

exports[`Text should render strike-through 1`] = `
.c0 {
  color: #404040;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  padding-bottom: 1rem;
  margin: 0;
  font-size: 0.9375em;
  line-height: 1.25rem;
}

@media (min-width:20em) {
  .c0 {
    font-size: 1em;
    line-height: 1.375rem;
  }
}

<p
  className="c0"
>
  This is text with 
  <del>
    some strike-through formatting
  </del>
  .
</p>
`;
